crates/kitsune2/src/factories/mem_peer_store.rs

```diff
@@ -0,0 +1,303 @@
+            }
+
+            if let Some(a) = self.store.get(&agent.agent) {
+                // If we already have a newer (or equal) one, abort.
+                if a.created_at >= agent.created_at {
+                    continue;
+                }
+            }
+
+            self.store.insert(agent.agent.clone(), agent);
+        }
+    }
+
+    pub fn get(&mut self, agent: AgentId) -> Option<Arc<AgentInfoSigned>> {
+        self.check_prune();
+
+        self.store.get(&agent).cloned()
+    }
+
+    pub fn get_all(&mut self) -> Vec<Arc<AgentInfoSigned>> {
+        self.check_prune();
+
+        self.store.values().cloned().collect()
+    }
+
+    pub fn get_overlapping_storage_arc(
+        &mut self,
+        arc: BasicArc,
+    ) -> Vec<Arc<AgentInfoSigned>> {
+        self.check_prune();
+
+        self.store
+            .values()
+            .filter_map(|info| {
+                if info.is_tombstone {
+                    return None;
+                }
+
+                if !arcs_overlap(arc, info.storage_arc) {
+                    return None;
+                }
+
+                Some(info.clone())
+            })
+            .collect()
+    }
+
+    pub fn get_near_location(
+        &mut self,
+        loc: u32,
+        limit: usize,
+    ) -> Vec<Arc<AgentInfoSigned>> {
+        self.check_prune();
+
+        let mut out: Vec<(u32, &Arc<AgentInfoSigned>)> = self
+            .store
+            .values()
+            .filter_map(|v| {
+                if !v.is_tombstone {
+                    if v.storage_arc.is_none() {
+                        // filter out zero arcs, they can't help us
+                        None
+                    } else {
+                        Some((calc_dist(loc, v.storage_arc), v))
+                    }
+                } else {
+                    None
+                }
+            })
+            .collect();
+
+        out.sort_by(|a, b| a.0.cmp(&b.0));
+
+        out.into_iter()
+            .take(limit)
+            .map(|(_, v)| v.clone())
+            .collect()
+    }
+}
+
+/// Get the min distance from a location to an arc in a wrapping u32 space.
+/// This function will only return 0 if the location is covered by the arc.
+/// This function will return u32::MAX if the arc is not set.
+fn calc_dist(loc: u32, arc: BasicArc) -> u32 {
+    match arc {
+        None => u32::MAX,
+        Some((arc_start, arc_end)) => {
+            let (d1, d2) = if arc_start > arc_end {
+                // this arc wraps around the end of u32::MAX
+
+                if loc >= arc_start || loc <= arc_end {
+                    return 0;
+                } else {
+                    (loc - arc_end, arc_start - loc)
+                }
+            } else {
+                // this arc does not wrap
+
+                if loc >= arc_start && loc <= arc_end {
+                    return 0;
+                } else if loc < arc_start {
+                    (arc_start - loc, u32::MAX - arc_end + loc + 1)
```

Comment:
Just to make sure I've got this understood correctly. The `u32::MAX - arc_end` is the distance from the end of the arc to the top of a u32, then `+ 1` wraps around to 0 and `+ loc` takes us back to our location?

I wonder if it might make this slightly clearer to organize the terms this way?

```suggestion
                    (arc_start - loc, (u32::MAX - arc_end + 1) + loc)
```

Reply:
Sooo... the actual reason is that `u32::MAX` is the final index, not the count. We actually want the count in this arithmetic, but since the count doesn't fit inside a u32, we have to be careful where we add it to make sure we don't overflow.

The parens, for me, are a lateral step, and don't help or hinder my understanding of this, so I'm happy to change it, but I'll wait for your response before implementing it.

Reply:
Understood :) This isn't important enough to me to change. Now I've had a chance to puzzle out the logic, I'm happy I could figure it out again written either way